File: test/e2e/Apps/BasicDotNetIsolated/SuspendResumeOrchestration.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Net;
using Grpc.Core;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask.Client;

namespace Microsoft.Azure.Durable.Tests.E2E;

public static class SuspendResumeOrchestration
{
    [Function("SuspendInstance")]
    public static async Task<HttpResponseData> Suspend(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client,
        string instanceId)
    {
        string suspendReason = "Suspending the instance for test.";
        try 
        {
            await client.SuspendInstanceAsync(instanceId, suspendReason);
            return req.CreateResponse(HttpStatusCode.OK);
        }
        catch (RpcException ex) 
        {
            var response = req.CreateResponse(HttpStatusCode.BadRequest);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync(ex.Message);
            return response;
        }
    }

    [Function("ResumeInstance")]
    public static async Task<HttpResponseData> Resume(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client,
        string instanceId)
    {
        string resumeReason = "Resuming the instance for test.";
        try 
        {
            await client.ResumeInstanceAsync(instanceId, resumeReason);
            return req.CreateResponse(HttpStatusCode.OK);
        }
        catch (RpcException ex) 
        {
            var response = req.CreateResponse(HttpStatusCode.BadRequest);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync(ex.Message);
            return response;
        }
    }
}

File: test/e2e/Apps/BasicPython/suspend_resume_orchestration.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.
#

import azure.functions as func
import azure.durable_functions as df

bp = df.Blueprint()


@bp.route(route="SuspendInstance", methods=["GET", "POST"])
@bp.durable_client_input(client_name="client")
async def suspend_instance(req: func.HttpRequest, client: df.DurableOrchestrationClient):
    instance_id = req.params.get("instanceId")
    suspend_reason = "Suspending the instance for test."
    try:
        await client.suspend(instance_id, suspend_reason)
        return func.HttpResponse(status_code=200)
    except Exception as ex:
        # Simulate RpcException handling and message
        response = func.HttpResponse(
            str(ex),
            status_code=400,
            mimetype="text/plain"
        )
        return response


@bp.route(route="ResumeInstance", methods=["GET", "POST"])
@bp.durable_client_input(client_name="client")
async def resume_instance(req: func.HttpRequest, client: df.DurableOrchestrationClient):
    instance_id = req.params.get("instanceId")
    resume_reason = "Resuming the instance for test."
    try:
        await client.resume(instance_id, resume_reason)
        return func.HttpResponse(status_code=200)
    except Exception as ex:
        # Simulate RpcException handling and message
        response = func.HttpResponse(
            str(ex),
            status_code=400,
            mimetype="text/plain"
        )
        return response
